trace/LogCorrelation.md

```diff
@@ -0,0 +1,63 @@
+# Log Correlation
+
+Log correlation is a feature that inserts information about the current span into log entries
+created by existing logging frameworks.  The feature can be used to add more context to log entries,
+filter log entries by trace ID, or show log entries as annotations on a trace.
+
+The design of a log correlation implementation depends heavily on the details of the particular
+logging framework that it supports.  Therefore, this document only covers the aspects of log
+correlation that could be shared across log correlation implementations for multiple languages and
+logging frameworks.  It doesn't cover how to hook into the logging framework.
+
+## Tracing data to include in log entries
+
+A log correlation implementation should insert the following pieces of tracing data from the current
+span context into each log entry:
+
+### Trace ID
+
+The trace ID of the current span.  See [Span.md#traceid](Span.md#traceid).
+
+### Span ID
+
+The span ID of the current span.  See [Span.md#spanid](Span.md#spanid).
+
+### Sampling Decision
+
+The sampling bit of the current span, as a boolean.  See
+[Span.md#supported-bits](Span.md#supported-bits).
+
+## String format for tracing data
+
+The logging framework may require the pieces of tracing data to be converted to strings.  In that
+case, the log correlation implementation should format the trace ID and span ID as lowercase base 16
+and format the sampling decision as "true" or "false".
+
+## Key names for tracing data
+
+Some logging frameworks allow the insertion of arbitrary key-value pairs into log entries.  When
+a log correlation implementation inserts tracing data by that method, the key names should be
+"opencensusTraceId", "opencensusSpanId", and "opencensusTraceSampled" by default.  The log
```

Comment:
These are very mouthful key names.

OTOH, underscore is often preferred over camel case in log labels. We might want to reconsider these keys.

Reply:
> These are very mouthful key names.

There was some discussion about short vs long key names in https://github.com/census-instrumentation/opencensus-java/pull/1371#discussion_r211440499.

I thought that it would be better to use long key names by default, to avoid conflicting with other tracing libraries' key names.  Users could then override the key names if they wanted to purposefully use the same key names for, say, Brave and OpenCensus.

> OTOH, underscore is often preferred over camel case in log labels. We might want to reconsider these keys.

I'm happy to change the key names if underscores are more common.  Where have you seen underscores used?

I saw camel case logging context key names in Brave, Wingtips, and Spring Cloud Sleuth and camel case tracing-related log entry fields in Stackdriver (https://cloud.google.com/logging/docs/agent/configuration#special_fields_in_structured_payloads).  The three libraries are all Java, though, so it could easily be Java specific.

Reply:
It could be a language-community preference. I saw it on Stackdriver Logging, zap and logrus.

https://github.com/uber-go/zap
https://github.com/sirupsen/logrus

Reply:
In the unusual case of conflict, users can always override the defaults.

Reply:
I think we should defer to language defaults. it would be weird to not use lower camel in most java logging formats (though not all). so we can make examples in a case format and caveat that appropriate use will occur when applied to a logging framework

Reply:
I think supporting per language name is not the best option:
1. Usually the log backends are not per language. So the backend has to understand all of these combinations.
2. I agree that size matters here so probably shorter is better (less data written).
3. I understand that for Java a key like "oc_trace_id" is probably not a good fit. For the moment we have only Java and Go here. Can we get some input from other languages?

Even if we have multiple keys we need to document them here for each language.

Reply:
I opened an issue for continuing the discussion: #195